### PR TITLE
add(scan): Add `ClearResults` and `DeleteKeys` gRPC methods

### LINKED ISSUE #8235: Add `ClearResults` and `DeleteKeys` gRPC methods
## Motivation

We want to be able to clear the results for a given viewing key and to delete the key and stop scanning for its transactions via RPC methods.

### PR DESCRIPTION
**Pull request by arya2**
## Motivation

We want to clear results and delete keys in zebra-scan via gRPC requests.

Closes #8235

#### Check before marking the PR as ready for review:
  - [x] Will the PR name make sense to users?
  - [x] Does the PR have a priority label?
  - [x] Have you added or updated tests?
  - [x] Is the documentation up to date?

##### For significant changes:
  - [x] Is there a summary in the CHANGELOG?
  - [x] Can these changes be split into multiple PRs?

_If a checkbox isn't relevant to the PR, mark it as done._

## Solution

- Adds the gRPC methods to the proto definition
- Implements the methods on `ScannerRPC` to call the `ScanService`

### Testing

These methods still need tests, but they're thin wrappers around scan service requests that do have tests.

## Review

Anyone can review.

### Reviewer Checklist

Check before approving the PR:
  - [ ] Does the PR scope match the ticket?
  - [ ] Are there enough tests to make sure it works? Do the tests cover the PR motivation?
  - [ ] Are all the PR blockers dealt with?
        PR blockers can be dealt with in new tickets or PRs.

_And check the PR Author checklist is complete._

## Follow Up Work

- Add tests to `zebra-grpc`

### CODE REVIEW

File: zebra-grpc/src/server.rs [hunk 1]
```diff
@@ -67,6 +67,60 @@ where
...
+    }
+
+    async fn delete_keys(
+        &self,
+        request: tonic::Request<DeleteKeysRequest>,
+    ) -> Result<Response<Empty>, Status> {
+        let keys = request.into_inner().keys;
+
+        if keys.is_empty() {
+            let msg = "must provide the keys for which to clear results";
```
Comment (arya2) [inline zebra-grpc/src/server.rs]:
```suggestion
            let msg = "must provide at least 1 key to delete";
```

File: zebra-grpc/src/server.rs [hunk 2]
```diff
@@ -67,6 +67,60 @@ where
...
     }
+
+    async fn clear_results(
+        &self,
+        request: tonic::Request<ClearResultsRequest>,
+    ) -> Result<Response<Empty>, Status> {
+        let keys = request.into_inner().keys;
+
+        if keys.is_empty() {
+            let msg = "must provide the keys for which to clear results";
```
Comment (arya2) [inline zebra-grpc/src/server.rs]:
```suggestion
            let msg = "must provide at least 1 key for which to clear results";
```

File: zebra-grpc/proto/scanner.proto [hunk 3]
```diff
@@ -7,10 +7,30 @@ message Empty {}
 service Scanner {
     // Get information about the scanner service.
     rpc GetInfo (Empty) returns (InfoReply);
+
+    // Clear results for a set of keys without removing the keys from the scanner.
+    // This request does not stop the scanner from scanning blocks for these keys, it
+    // only clears past results.
+    rpc ClearResults(ClearResultsRequest) returns (Empty);
+
+    // Deletes a set of keys and their results from the scanner
```
Comment (oxarbitrage) [inline zebra-grpc/proto/scanner.proto]:
```suggestion
    // Deletes a set of keys and their results from the scanner.
    // This request stop the scanner from scanning blocks for the these keys.
```

File: zebra-grpc/src/server.rs [hunk 4]
```diff
@@ -67,6 +67,60 @@ where
...
+            .and_then(|service| service.call(ScanServiceRequest::ClearResults(keys)))
+            .await
+            .map_err(|err| Status::unknown(format!("scan service returned error: {err}")))?
+        else {
+            return Err(Status::unknown(
+                "scan service returned an unexpected response",
+            ));
+        };
+
+        Ok(Response::new(Empty {}))
```
Comment (oxarbitrage) [inline zebra-grpc/src/server.rs:95]: Do we want to return the removed results ?

File: zebra-grpc/src/server.rs [hunk 5]
```diff
@@ -67,6 +67,60 @@ where
...
+            .and_then(|service| service.call(ScanServiceRequest::DeleteKeys(keys)))
+            .await
+            .map_err(|err| Status::unknown(format!("scan service returned error: {err}")))?
+        else {
+            return Err(Status::unknown(
+                "scan service returned an unexpected response",
+            ));
+        };
+
+        Ok(Response::new(Empty {}))
```
Comment (oxarbitrage) [inline zebra-grpc/src/server.rs:122]: Do we want to return removed keys?

**Review by oxarbitrage — DISMISSED** [review]
Looks good to me but i think we might want tests. We can do all the grpc testing (including snapshots) for all grpc in a separated ticket or add tests here so the other calls can follow.

Can you make a manual test and post the output ? A manual test should be to start the server and run these calls, make sure they reply as expected.

File: zebra-grpc/src/server.rs (hunk 4 above)
Comment (arya2) [reply zebra-grpc/src/server.rs:95]: I think we should later accept an optional height range or a stop height for clearing only some of the results, but only return an empty response if it's successful or an error from this RPC, assuming that if they're clearing the results, they already have them or are no longer interested in them.

File: zebra-grpc/src/server.rs (hunk 5 above)
Comment (arya2) [reply zebra-grpc/src/server.rs:122]: I don't think it's necessary, it would just be sending back the keys that the client provided.

**Comment by arya2** [comment]
> Looks good to me but i think we might want tests. We can do all the grpc testing (including snapshots) for all grpc in a separated ticket or add tests here so the other calls can follow.

I'll open an issue for adding tests for the gRPC methods, I think it's okay to add them later since the RPC methods are thin wrappers around the scan service calls.

> Can you make a manual test and post the output ? A manual test should be to start the server and run these calls, make sure they reply as expected.

Yep, I'll start the server from zebrad in a follow-up PR targeting this branch for the manual test.

File: zebra-grpc/src/server.rs (hunk 4 above)
Comment (oxarbitrage) [reply zebra-grpc/src/server.rs:95]: ok, that makes sense, thank you.

**Comment by arya2** [comment]
Manual test of these methods:

```console

~/github/zebra$ $HOME/go/bin/grpcurl -plaintext -import-path ./zebra-grpc/proto -proto zebra-grpc/proto/scanner.proto -d '{"keys": [""]}' '[::1]:50051' scanner.Scanner/ClearResults
{}
~/github/zebra$ $HOME/go/bin/grpcurl -plaintext -import-path ./zebra-grpc/proto -proto zebra-grpc/proto/scanner.proto -d '{"keys": [""]}' '[::1]:50051' scanner.Scanner/DeleteKeys
{}

```

It seems strange to have a successful response when the provided keys were not registered in the scanner, we may want to return an error in that case later.

**Review by oxarbitrage — APPROVED** [review]

**Pull request merged**